src/core/service.c

```diff
@@ -252,6 +254,30 @@ static void service_start_watchdog(Service *s) {
                 log_unit_warning_errno(UNIT(s), r, "Failed to install watchdog timer: %m");
 }
 
+static void service_extend_timeout(Service *s) {
+        int r;
+
+        assert(s);
+
+        if (IN_SET(s->state, SERVICE_START, SERVICE_START_POST,
+                    SERVICE_STOP, SERVICE_STOP_SIGABRT, SERVICE_STOP_SIGTERM, SERVICE_STOP_SIGKILL, SERVICE_STOP_POST,
+                    SERVICE_FINAL_SIGTERM, SERVICE_FINAL_SIGKILL)) {
+                usec_t watchdog_usec, new_timeout;
+
+                watchdog_usec = service_get_watchdog_usec(s);
+                new_timeout = usec_add(now(CLOCK_MONOTONIC), watchdog_usec);
+                r = service_arm_timer(s, new_timeout);
```

Comment:
hmm, this way as soon as we get EXTEND_TIMEOUT=1 we completely forget about any timeout set with TimeoutStartSec= and TimeoutStopSec=. I find that quite surprising I think, no? 

Instead of calling service_arm_timer() here, I'd instead do something like this:

```
if (s->timer_event_source) {
        uint64_t current, extended;
        r = sd_event_source_get_time(s->timer_event_source, &current);
        if (r < 0)
                …
        extended = usec_add(now(CLOCK_MONOTONIC), service_get_watchdog_usec(s));
        if (extended > current) {
                r = sd_event_source_set_time(s->timer_event_source, extended),
                if (r < 0) 
                        …
         }
}
```

You get the idea: we bump the timeout only if we'd actually increase it, and otherwise just leave in what's already set.

(and similar for the watchdog timeout event source)

Reply:
Forgot to update documentation. Will look tomorrow. Need to do "similar for watchdog timeout event source too" (with tests)